auth.go

```diff
@@ -28,31 +32,85 @@ func pathLogin(b *mesosBackend) *framework.Path {
 
 // pathLogin (the method) is the "login" path request handler.
 func (b *mesosBackend) pathLogin(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-
 	taskID := d.Get("task-id").(string)
-	if len(taskID) < 1 {
+	if !b.verifyTaskExists(taskID) {
+		return nil, logical.ErrPermissionDenied
+	}
+
+	prefix, err := taskIDPrefix(taskID)
+	if err != nil {
 		return nil, logical.ErrPermissionDenied
 	}
 
-	b.Logger().Info("LOGIN", "task-id", taskID, "RemoteAddr", req.Connection.RemoteAddr)
+	b.Logger().Info("LOGIN", "task-id", taskID, "prefix", prefix, "RemoteAddr", req.Connection.RemoteAddr)
+
+	policies, err := getTaskPolicies(ctx, req.Storage, prefix)
+	if err != nil {
+		return nil, err
+	}
 
-	// TODO: Validate the task-id and look up the associated list of policies.
 	// TODO: Make the renewal period configurable?
 	return &logical.Response{
 		Auth: &logical.Auth{
-			Policies: []string{},
+			Policies: policies,
 			Period:   10 * time.Minute,
 			LeaseOptions: logical.LeaseOptions{
 				Renewable: true,
 			},
+			// Stash task-id so we can check it again for renewals.
+			InternalData: jsonobj{"task-id": taskID},
 		},
 	}, nil
 }
 
 // authRenew is the renew callback for tokens created by this plugin.
 func (b *mesosBackend) authRenew(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	b.Logger().Info("RENEW", "req", fmt.Sprintf("%#v", req))
-	// For an unconditional renewal, we only need to return the Auth struct
+	b.Logger().Info("RENEW", "auth", fmt.Sprintf("%#v", req.Auth))
+
+	taskID := req.Auth.InternalData["task-id"].(string)
+	if !b.verifyTaskExists(taskID) {
+		return nil, fmt.Errorf("task %s not found during renewal", taskID)
+	}
+
+	// For a standard periodic renewal, we only need to return the Auth struct
 	// we're given in the request.
 	return &logical.Response{Auth: req.Auth}, nil
 }
+
+// verifyTaskExists checks that a taskID is valid and identifies an existing
+// task.
+func (b *mesosBackend) verifyTaskExists(taskID string) bool {
+	if taskID == "" {
+		return false
+	}
+
+	b.Logger().Debug("TODO: Check task in mesos.")
+	// TODO: Verify that the task exists.
+	return temporarySetOfExistingTasks[taskID]
+}
+
+// getTaskPolicies fetches the policies for a taskID prefix.
+func getTaskPolicies(ctx context.Context, storage logical.Storage, taskPrefix string) ([]string, error) {
+	se, err := storage.Get(ctx, tpKey(taskPrefix))
+	if err != nil {
+		// noqa: (Not actually a tag that does anything, sadly.)
+		return nil, err
+	}
+	if se == nil {
+		return nil, logical.ErrPermissionDenied
+	}
+
+	var tp taskPolicies
+	err = se.DecodeJSON(&tp)
+	return tp.Policies, err
+}
+
+// taskIDPrefix extracts the prefix from a taskID.
+func taskIDPrefix(taskID string) (string, error) {
+	idx := strings.LastIndex(taskID, ".")
```

Comment:
This is quite a specific kind of "prefix" so not sure I like it being called `task-id-prefix`... but that's probably something for a later PR.

Reply:
Yeah. Naming is hard, and I'm still not entirely sure whether we want to support multiple frameworks, different kinds of frameworks, etc.

For now I'm assuming "the kind of Mesos/Marathon we have in our DC/OS clusters", but we may want to revisit that later.